metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,77 +1,139 @@
-   *
-   * For every test suite a.MunitTestSuite BSP returns 2 symbols. However, only one of them
-   * has location in source code and it's useful.
-   * That's why FullyQualifiedClassName is used as a key.
-   * "a/MunitTestSuite., a.MunitTestSuite"
-   * "a/MunitTestSuite#, a.MunitTestSuite"
+   * Update test cases for given path.
+   * Only consider 'relevant' files which have had already discovered test suite.
+   * Metals have to update them because they can be visible to the user via:
+   * 1. Test Explorer view can be opened and tree view is visible
+   * 2. test suite's file can be opened
    */
-  private val cachedTestSuites =
-    mutable.Map[FullyQualifiedClassName, TestEntry]()
+  def onChange(docs: TextDocuments, file: AbsolutePath): Unit = {
+    if (isEnabled && index.contains(file)) {
+      if (docs.documents.nonEmpty) {
+        val doc = docs.documents.head
+        index.updateDoc(file, doc)
+      }
+      refreshTestCases(file)
+    }
+  }
 
-  val refreshTestSuites = new BatchedFunction[Unit, Unit](_ =>
-    CancelableFuture(doRefreshTestSuites())
-  )
+  /**
+   * Check if opened file contains test suite and update test cases if yes.
+   */
+  def didOpen(file: AbsolutePath): Future[Unit] =
+    if (isEnabled && index.contains(file)) Future(refreshTestCases(file))
+    else Future.unit
 
   /**
-   * Retrieves cached test suites.
+   * Retrieves all cached test suites. Useful for tests.
    */
-  def findTestSuites(): java.util.List[TestSuiteDiscoveryResult] = {
-    cachedTestSuites.values
-      .groupBy(_.buildTarget)
-      .toSeq
+  def getTestSuites(): java.util.List[BuildTargetUpdate] = {
+    index.suites
       .map { case (buildTarget, entries) =>
-        val grouped = groupTestsByPackage(entries.toList)
-        TestSuiteDiscoveryResult(
-          buildTarget.getDisplayName(),
-          buildTarget.getId().getUri(),
-          grouped.asJava
-        )
+        BuildTargetUpdate(buildTarget, entries.map(_.testClass).toSeq)
       }
+      .toSeq
       .asJava
   }
 
+  /**
+   * Retrieve test cases for a given file. Even an empty list
+   * because it can mean that all testcases were deleted.
+   */
+  def getTestCases(path: AbsolutePath): Seq[BuildTargetUpdate] = {
+    val buildTargetUpdateOpt =
+      for {
+        metadata <- index.getMetadata(path)
+        buildTarget <- metadata.entries.headOption.map(_.buildTarget)
+      } yield {
+        val events = findTestCases(path)
+        BuildTargetUpdate(buildTarget, events)
+      }
+    buildTargetUpdateOpt.toSeq
+  }
+
+  private def refreshTestCases(path: AbsolutePath): Unit = {
+    val buildTargetUpdateOpt = getTestCases(path)
+    buildTargetUpdateOpt.foreach { update =>
+      client.metalsExecuteClientCommand(
+        ClientCommands.UpdateTestExplorer.toExecuteCommandParams(
+          update
+        )
+      )
+    }
+  }
+
+  /**
+   * Searches for test cases for a given path.
+   */
+  private def findTestCases(
```

Comment:
Test cases are still not supported to be run individually. Would it make sense to make this PR first work with push, but exclude test cases? So that we don't put everything into one big PR?

Reply:
Yes, they aren't yet supported. But it will be really helpful if it'll be merged now.

Reply:
I realized this after I wrote the comment, cool!